keras/utils/dataset_utils.py

```diff
@@ -19,9 +19,395 @@
+                                                to validate the shape of rest
+                                                of the samples.
+                                                Defaults to True.
+      start_time (float): the start time of the dataset iteration. this is
+                          used only if `data_size_warning_flag` is set to true.
+
+  Raises:
+      ValueError: - If the dataset is empty.
+                  - If `ensure_shape_similarity` is set to True and the
+                    shape of the first sample is not equal to the shape of
+                    atleast one of the rest of the samples.
+
+  yields:
+      data_sample: A tuple/list of numpy arrays.
+  """
+  try:
+    dataset_iterator = iter(dataset_iterator)
+    first_sample = next(dataset_iterator)
+    if isinstance(first_sample,(tf.Tensor,np.ndarray)):
+      first_sample_shape = np.array(first_sample).shape
+    else:
+      first_sample_shape = None
+      ensure_shape_similarity  = False
+    yield first_sample
+  except StopIteration:
+    raise ValueError('Received an empty Dataset. `dataset` must '
+                     'be a non-empty list/tuple of numpy.ndarrays '
+                     'or tf.data.Dataset objects.')
+
+  for i,sample in enumerate(dataset_iterator):
+    if ensure_shape_similarity:
+      if first_sample_shape != np.array(sample).shape:
+        raise ValueError('All `dataset` samples must have same shape, '
+                        f'Expected shape: {np.array(first_sample).shape} '
+                        f'Received shape: {np.array(sample).shape} at index '
+                        f'{i}.')
+    if data_size_warning_flag:
+      if i % 10 == 0:
+        cur_time = time.time()
+        # warns user if the dataset is too large to iterate within 10s
+        if int(cur_time - start_time) > 10 and data_size_warning_flag:
+          warnings.warn('The dataset is taking longer than 10 seconds to '
+                        'iterate. This may be due to the size of the dataset. '
+                        'Please consider using a smaller dataset'
+                        '(e.g. < 10,000 samples). \nTo hide this '
+                        'warning message, set `data_size_warning_flag=False`.',
+                        category=ResourceWarning,
+                        source='split_dataset')
+          data_size_warning_flag = False
+    yield sample
+
+def _restore_dataset_from_list(dataset_as_list,dataset_type_spec,
+                               original_dataset):
+  """Restore the dataset from the list of arrays."""
+  if dataset_type_spec  in [tuple,list]:
+    return tuple(np.array(sample) for sample in zip(*dataset_as_list))
+  elif dataset_type_spec == tf.data.Dataset:
+    if type(original_dataset.element_spec) is dict:
+      restored_dataset = dict()
+      for d in dataset_as_list:
+        for k, v in d.items():
+          if k not in restored_dataset:
+            restored_dataset[k] = [np.array(v)]
```

Comment:
Why `np.array`? Would that work with string types?

Reply:
Thanks for pointing it out. `np.array()` will lead to unnecessary type conversion. I removed it